tests/cpp/test_abstract_tensor.cpp

```diff
@@ -712,4 +717,72 @@ TEST_F(AbstractTensorTest, UnzipBroadcasting) {
   EXPECT_EQ(ub[1], expect1);
 }
 
+TEST_F(AbstractTensorTest, TestApplyScheduling) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeSymbolicTensor(1);
+  auto tv1 = makeSymbolicTensor(2);
+  auto tv2 = makeSymbolicTensor(2);
+  fusion.addInput(tv0);
+  fusion.addInput(tv1);
+  fusion.addInput(tv2);
+
+  // This fusion does not have a reference tensor
+  auto tv3 = set(tv0);
```

Comment:
I'm currently trying to understand this. If I remove this `set` the test passes. Otherwise, it fails during lowering (the transform checks pass):
> C++ exception with description "alloc_ind != nullptr INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/index_compute.cpp":1818, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. Couldn't find allocation mapping for T1_g[ iS65{( ceilDiv(( ( (( (( getMetaData(T0) )).logical_size ))[0] ) * ( (( (( getMetaData(T1) )).logical_size ))[1] ) ), 128) )}, iS66{128} ] dim: 1 id: iS53{( (( (( getMetaData(T1) )).logical_size ))[1] )}

Reply:
I think this failure might show why #2459 is preferable to forwarding through merges around missing axes. `tv3` here is 1D and scheduled with split(128), while the other tensors get merged then split. Probably all tensors need to be scheduling-broadcast to 3D in order to get the indexing right.

Reply:
However, we don't hit this error with a single set+broadcast+add kernel, so it's probably just another bug.

Reply:
What does the final fusion math look like?

Reply:
```
Inputs:
  T0_g[ iblockIdx.x17{( ceilDiv(i0, 128) )}, ithreadIdx.x18{128} ], float
  T1_g[ iblockIdx.x28{( ceilDiv(( i2 * i3 ), 128) )}, ithreadIdx.x29{128} ], float
  T2_g[ iblockIdx.x43{( ceilDiv(( i4 * i5 ), 128) )}, ithreadIdx.x44{128} ], float
Outputs:
  T5_g[ iblockIdx.x34{( ceilDiv(( i0 * i3 ), 128) )}, ithreadIdx.x35{128} ] ca_pos( 2 ) produce_pos( 2 ), float
  T7_g[ iblockIdx.x49{( ceilDiv(( i4 * i0 ), 128) )}, ithreadIdx.x50{128} ] ca_pos( 2 ) produce_pos( 2 ), float

%kernel_math {
T3_l[ iblockIdx.x19{( ceilDiv(i0, 128) )}, ithreadIdx.x20{128} ] ca_pos( 2 )
   = Set( T0_g[ iblockIdx.x17{( ceilDiv(i0, 128) )}, ithreadIdx.x18{128} ], cache_op=Streaming )
T4_l[ iblockIdx.x31{( ceilDiv(( i0 * 1 ), 128) )}, ithreadIdx.x32{128} ] ca_pos( 2 ) produce_pos( 2 )
   = broadcast( T3_l[ iblockIdx.x19{( ceilDiv(i0, 128) )}, ithreadIdx.x20{128} ] ca_pos( 2 ) )
T5_g[ iblockIdx.x34{( ceilDiv(( i0 * i3 ), 128) )}, ithreadIdx.x35{128} ] ca_pos( 2 ) produce_pos( 2 )
   = T4_l[ iblockIdx.x31{( ceilDiv(( i0 * 1 ), 128) )}, ithreadIdx.x32{128} ] ca_pos( 2 ) produce_pos( 2 )
   + T1_g[ iblockIdx.x28{( ceilDiv(( i2 * i3 ), 128) )}, ithreadIdx.x29{128} ];
T6_l[ iblockIdx.x46{( ceilDiv(( 1 * i0 ), 128) )}, ithreadIdx.x47{128} ] ca_pos( 2 ) produce_pos( 2 )
   = broadcast( T3_l[ iblockIdx.x19{( ceilDiv(i0, 128) )}, ithreadIdx.x20{128} ] ca_pos( 2 ) )
T7_g[ iblockIdx.x49{( ceilDiv(( i4 * i0 ), 128) )}, ithreadIdx.x50{128} ] ca_pos( 2 ) produce_pos( 2 )
   = T6_l[ iblockIdx.x46{( ceilDiv(( 1 * i0 ), 128) )}, ithreadIdx.x47{128} ] ca_pos( 2 ) produce_pos( 2 )
   + T2_g[ iblockIdx.x43{( ceilDiv(( i4 * i5 ), 128) )}, ithreadIdx.x44{128} ];
}
```

Reply:
I'm actually surprised it works when the set isn't there. Looks like all tensors are inlined at position 2, but none of them should be concrete. 

Could you also show the math output without the set?

Reply:
Sure. Without the set:
```
Inputs:
  T0_g[ iblockIdx.x16{( ceilDiv(i0, 128) )}, ithreadIdx.x17{128} ], float
  T1_g[ iblockIdx.x25{( ceilDiv(( i2 * i3 ), 128) )}, ithreadIdx.x26{128} ], float
  T2_g[ iblockIdx.x40{( ceilDiv(( i4 * i5 ), 128) )}, ithreadIdx.x41{128} ], float
Outputs:
  T4_g[ iblockIdx.x31{( ceilDiv(( i0 * i3 ), 128) )}, ithreadIdx.x32{128} ] ca_pos( 2 ) produce_pos( 2 ), float
  T6_g[ iblockIdx.x46{( ceilDiv(( i4 * i0 ), 128) )}, ithreadIdx.x47{128} ] ca_pos( 2 ) produce_pos( 2 ), float

%kernel_math {
T3_l[ iblockIdx.x28{( ceilDiv(( i0 * 1 ), 128) )}, ithreadIdx.x29{128} ] ca_pos( 2 )
   = broadcast( T0_g[ iblockIdx.x16{( ceilDiv(i0, 128) )}, ithreadIdx.x17{128} ] )
T4_g[ iblockIdx.x31{( ceilDiv(( i0 * i3 ), 128) )}, ithreadIdx.x32{128} ] ca_pos( 2 ) produce_pos( 2 )
   = T3_l[ iblockIdx.x28{( ceilDiv(( i0 * 1 ), 128) )}, ithreadIdx.x29{128} ] ca_pos( 2 )
   + T1_g[ iblockIdx.x25{( ceilDiv(( i2 * i3 ), 128) )}, ithreadIdx.x26{128} ];
T5_l[ iblockIdx.x43{( ceilDiv(( 1 * i0 ), 128) )}, ithreadIdx.x44{128} ] ca_pos( 2 )
   = broadcast( T0_g[ iblockIdx.x16{( ceilDiv(i0, 128) )}, ithreadIdx.x17{128} ] )
T6_g[ iblockIdx.x46{( ceilDiv(( i4 * i0 ), 128) )}, ithreadIdx.x47{128} ] ca_pos( 2 ) produce_pos( 2 )
   = T5_l[ iblockIdx.x43{( ceilDiv(( 1 * i0 ), 128) )}, ithreadIdx.x44{128} ] ca_pos( 2 )
   + T2_g[ iblockIdx.x40{( ceilDiv(( i4 * i5 ), 128) )}, ithreadIdx.x41{128} ];
}
```

Reply:
I don't understand how this actually works. There's no single domain that actually can work as a concrete domain, yet all tensors are fully inlined. 

Does compute-sanitizer show anything?

I'd also be interested in the generated kernel *without* parallelization and see what the outer loop looked like. That should tell us what domain is used as the concrete domain.

Reply:
No errors in compute-sanitizer. Without parallelization, we have two separate loop nests:
```c++
__global__ void nvfuser_none_f0_c0_r0_g0(Tensor<float, 1, 1> T0, Tensor<float, 2, 2> T1, Tensor<float, 2, 2> T2, Tensor<float, 2, 2> T4, Tensor<float, 2, 2> T6) {
  NVFUSER_DEFINE_MAGIC_ZERO;
  nvfuser_index_t i0;
  i0 = T2.logical_size[0LL] * T0.logical_size[0LL];
  nvfuser_index_t i1;
  i1 = ceilDiv(i0, 128LL);
  nvfuser_index_t i2;
  i2 = T0.logical_size[0LL] * T1.logical_size[1LL];
  nvfuser_index_t i3;
  i3 = ceilDiv(i2, 128LL);
  nvfuser_index_t i4;
  i4 = T1.logical_size[1LL] * T0.logical_size[0LL];
  #pragma unroll 1
  for(nvfuser_index_t i5 = 0; i5 < i1; ++i5) {
    nvfuser_index_t i6;
    i6 = 128LL * i5;
    #pragma unroll
    for(nvfuser_index_t i7 = 0; i7 < 128LL; ++i7) {
      nvfuser_index_t i8;
      i8 = i6 + (i7 + nvfuser_zero);
      nvfuser_index_t i9;
      i9 = i8 % T0.logical_size[0LL];
      float T5[1LL];
      T5[0LL]
         = T0[(T0.alloc_stride[0LL] * i9)];
      if ((i8 < i0)) {
        T6[i8]
          = T5[0LL]
          + T2[((T2.alloc_stride[0LL] * (i8 / T0.logical_size[0LL])) + (T2.alloc_stride[1LL] * i9))];
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  }
  #pragma unroll 1
  for(nvfuser_index_t i10 = 0; i10 < i3; ++i10) {
    nvfuser_index_t i11;
    i11 = 128LL * i10;
    #pragma unroll
    for(nvfuser_index_t i12 = 0; i12 < 128LL; ++i12) {
      nvfuser_index_t i13;
      i13 = i11 + (i12 + nvfuser_zero);
      nvfuser_index_t i14;
      i14 = i13 / T1.logical_size[1LL];
      float T3[1LL];
      if ((i13 < i4)) {
        T3[0LL]
           = T0[(T0.alloc_stride[0LL] * i14)];
      }
      if ((i13 < i2)) {
        T4[i13]
          = T3[0LL]
          + T1[((T1.alloc_stride[0LL] * i14) + (T1.alloc_stride[1LL] * (i13 % T1.logical_size[1LL])))];
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  }
}
```

Reply:
Oh, I now see what's going on. Without the set, there are essentially two disjoint groups of tensors. The only common tensor is t0, but since it's an input, it doesn't matter for inlining.

With the set, however, an intermediate tensor, t3, is added, and that's another common tensor, and since it's not an input, inlining does matter for t3. Since t3 is used by both of the two groups of tensors, it effectively forces the two groups to be inlined together, and that's when the problem of missing concrete domains happens.

This pattern is similar to the fusion in this [test.](https://github.com/NVIDIA/Fuser/blob/main/tests/cpp/test_id_model.cpp#L1225). The fusion doesn't work with the current lowering. Specifically, an incorrect loop and indices would be generated. The loop promotion analysis should capture all the necessary information to fix the problem, but I haven't yet looked into the issue very closely.

I don't think the loop promotion analysis, at least at this point, would work with this fusion with the set. The analysis relies on the broadcast forwarding pattern to infer the correct concrete domain of each loop group, but in this case there's no broadcast.

I think it's possible to extend the loop promotion analysis, but I don't feel that's the right approach.